scripts/bootloader/zb_ota/package.py

```diff
@@ -0,0 +1,446 @@
+      |- is_debug: {18}
+
+""".format(index,
+        type_strs[hex_type],
+        img.bin_file,
+        img.dat_file,
+        CommandTypes(cmd.op_code).name,
+        signature_type,
+        signature_hex,
+        cmd.init.fw_version,
+        cmd.init.hw_version,
+        DFUType(cmd.init.type).name,
+        cmd.init.app_size,
+        HashTypes(cmd.init.hash.hash_type).name,
+        binascii.hexlify(cmd.init.hash.hash),
+        )
+
+        return s
+
+    def __str__(self):
+
+        imgs = ""
+        i = 0
+        if self.manifest.application:
+            imgs = imgs + self.image_str(i, HexType.APPLICATION, self.manifest.application)
+            i = i + 1
+
+        s = """
+DFU Package: <{0}>:
+|
+|- Image count: {1}
+""".format(self.zip_file, i)
+
+        s = s + imgs
+        return s
+
+    def generate_package(self, filename, preserve_work_dir=False):
+        """
+        Generates a Nordic DFU package. The package is a zip file containing firmware(s) and metadata required
+        for Nordic DFU applications to perform DFU onn nRF5X devices.
+
+        :param str filename: Filename for generated package.
+        :param bool preserve_work_dir: True to preserve the temporary working directory.
+        Useful for debugging of a package, and if the user wants to look at the generated package without having to
+        unzip it.
+        :return: None
+        """
+        self.zip_file = filename
+        self.work_dir = self.__create_temp_workspace()
+
+        for key, firmware_data in self.firmwares_data.items():
+
+            # Normalize the firmware file and store it in the work directory
+            firmware_data[FirmwareKeys.BIN_FILENAME] = \
+                Package.normalize_firmware_to_bin(self.work_dir, firmware_data[FirmwareKeys.FIRMWARE_FILENAME])
+
+            # Calculate the hash for the .bin file located in the work directory
+            bin_file_path = os.path.join(self.work_dir, firmware_data[FirmwareKeys.BIN_FILENAME])
+            firmware_hash = Package.calculate_sha256_hash(bin_file_path)
+            bin_length = int(Package.calculate_file_size(bin_file_path))
+
+            app_size = 0
+            if key in [HexType.APPLICATION, HexType.EXTERNAL_APPLICATION]:
+                app_size = bin_length
+
+            init_packet = InitPacketPB(
+                            from_bytes = None,
+                            hash_bytes=firmware_hash,
+                            hash_type=HashTypes.SHA256,
+                            dfu_type=HexTypeToInitPacketFwTypemap[key],
+                            fw_version=firmware_data[FirmwareKeys.INIT_PACKET_DATA][PacketField.FW_VERSION],
+                            hw_version=firmware_data[FirmwareKeys.INIT_PACKET_DATA][PacketField.HW_VERSION],
+                            app_size=app_size)
+
+            # Store the .dat file in the work directory
+            init_packet_filename = firmware_data[FirmwareKeys.BIN_FILENAME].replace(".bin", ".dat")
```

Comment:
In theory there can be `.bin` string inside filename (not in the end of a file) and these occurrences will be also replaced

Reply:
fixed